src/server/game/Movement/MotionMaster.cpp

```diff
@@ -302,6 +302,23 @@ void MotionMaster::MovePoint(uint32 id, float x, float y, float z, bool generate
     }
 }
 
+void MotionMaster::MoveCloserAndStop(uint32 id, Unit* target, float distance)
+{
+    float distanceToTravel = _owner->GetExactDist2d(target) - distance;
+    if (distanceToTravel > 0.0)
+    {
+        float angle = _owner->GetAngle(target);
+        float destx = _owner->GetPositionX() + distanceToTravel * std::cos(angle);
+        float desty = _owner->GetPositionY() + distanceToTravel * std::sin(angle);
+        MovePoint(id, destx, desty, target->GetPositionZ());
+    }
+    else 
+    {
+        // we are already close enough. We just need to turn toward the target without changing position.
+        _owner->SetFacingToObject(target);
```

Comment:
What happens when the target moves meanwhile (in or against the direction the target is moving)?